libsolidity/formal/Z3CHCSmtLib2Interface.cpp

```diff
@@ -47,34 +62,56 @@ CHCSolverInterface::QueryResult Z3CHCSmtLib2Interface::query(smtutil::Expression
 {
 	setupSmtCallback(true);
 	std::string query = dumpQuery(_block);
-	std::string response = querySolver(query);
-	// NOTE: Our internal semantics is UNSAT -> SAFE and SAT -> UNSAFE, which corresponds to usual SMT-based model checking
-	// However, with CHC solvers, the meaning is flipped, UNSAT -> UNSAFE and SAT -> SAFE.
-	// So we have to flip the answer.
-	if (boost::starts_with(response, "unsat"))
+	try
 	{
-		// Repeat the query with preprocessing disabled, to get the full proof
-		setupSmtCallback(false);
-		query = "(set-option :produce-proofs true)" + query + "\n(get-proof)";
-		response = querySolver(query);
-		setupSmtCallback(true);
-		if (!boost::starts_with(response, "unsat"))
-			return {CheckResult::SATISFIABLE, Expression(true), {}};
-		return {CheckResult::SATISFIABLE, Expression(true), graphFromZ3Answer(response)};
-	}
+#ifdef EMSCRIPTEN_BUILD
+		z3::set_param("fp.xform.slice", true);
+		z3::set_param("fp.xform.inline_linear", true);
+		z3::set_param("fp.xform.inline_eager", true);
+		std::string response = [&](){ z3::context context; return Z3_eval_smtlib2_string(context, query.c_str()); }();
```

Comment:
Why not just this or similar instead of the lambda?
```suggestion
		std::string response = Z3_eval_smtlib2_string(z3::context{}, query.c_str());
```

Reply:
Thank you for the suggestion, I haven't thought of that!
This should indeed work and is much nicer!
Please, check the updated code.